### PR TITLE
BGL bugfix

### PR DESCRIPTION
**Pull request by gdamiand**
* Bug fix in add_vertex_and_face_to_border in  Euler_operations.h
* Bug fix in test_Euler_operations.cpp: use openmesh (and not surface mesh) if we want to test openmesh.

### CODE REVIEW

**Comment by lrineau** [comment]
@sloriot @janetournois @afabri Please review the fixes.

**Comment by lrineau** [comment]
Should not it be applied to CGAL-4.8 as well?

**Comment by gdamiand** [comment]
It could be; however I grep and it seems the function add_vertex_and_face_to_border is never used (except in test) thus this issue seems not crucial...

**Comment by lrineau** [comment]
There are compilation errors:

* https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-10/BGL/TestReport_lrineau_Fedora.gz
```
/usr/bin/c++   -DCGAL_TEST_SUITE -DCGAL_USE_GMP -DCGAL_USE_MPFR -DCGAL_USE_OPENMESH -DNOMINMAX -D_USE_MATH_DEFINES -I/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/../../include -I/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL -I/home/cgal_tester/build/src/cmake/platforms/Fedora/include -I/mnt/testsuite/include  -Wall -frounding-math   -o CMakeFiles/test_Euler_operations.dir/test_Euler_operations.cpp.o -c /home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp
/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp: In instantiation of 'void join_face_test() [with T = OpenMesh::PolyMesh_ArrayKernelT<>]':
/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp:364:24:   required from 'void test_Euler_operations() [with Graph = OpenMesh::PolyMesh_ArrayKernelT<>]'
/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp:386:32:   required from here
/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp:27:49: error: 'halfedges_around_face' was not declared in this scope
   boost::tie(begin, end) = halfedges_around_face(halfedge(f.f1, f.m), f.m);
                                                 ^
/home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp:27:49: note: suggested alternative:
In file included from /mnt/testsuite/include/CGAL/boost/graph/graph_traits_HalfedgeDS.h:37:0,
                 from /mnt/testsuite/include/CGAL/boost/graph/graph_traits_Polyhedron_3.h:23,
                 from /mnt/testsuite/include/CGAL/Polyhedron_3.h:1541,
                 from /mnt/testsuite/include/CGAL/IO/Polyhedron_iostream.h:24,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Prefix.h:14,
                 from /home/cgal_tester/build/src/cmake/platforms/Fedora/test/BGL/test_Euler_operations.cpp:2:
/mnt/testsuite/include/CGAL/boost/graph/iterator.h:869:1: note:   'CGAL::halfedges_around_face'
 halfedges_around_face(typename boost::graph_traits<Graph>::halfedge_descriptor h, const Graph& g)
 ^
```
and other errors that may or may not be consequences of the first error.

* https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-10/BGL/TestReport_lrineau_x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8.gz
```
/usr/local/packages/llvm-trunk/bin/clang++   -DCGAL_TEST_SUITE -DCGAL_USE_GMP -DCGAL_USE_GMPXX -DCGAL_USE_MPFR -DCGAL_USE_OPENMESH -DNOMINMAX -D_USE_MATH_DEFINES -I/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL/../../include -I/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL -I/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/include -I/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/include  -fno-strict-aliasing -Wall -fsanitize=address -DDONT_USE_BOOST_PROGRAM_OPTIONS -O2 -fno-omit-frame-pointer   -o CMakeFiles/test_Euler_operations.dir/test_Euler_operations.cpp.o -c /no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL/test_Euler_operations.cpp
/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL/test_Euler_operations.cpp:27:28: error: use of undeclared identifier 'halfedges_around_face'
  boost::tie(begin, end) = halfedges_around_face(halfedge(f.f1, f.m), f.m);
                           ^
/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL/test_Euler_operations.cpp:364:3: note: in instantiation of function template specialization 'join_face_test<OpenMesh::PolyMesh_ArrayKernelT<OpenMesh::DefaultTraits> >' requested here
  join_face_test<Graph>();
  ^
/no-raid/home/lrineau/CGAL/CGAL-4.9-Ic-10/cmake/platforms/x86-64_Linux-Fedora19_llvm-clang-with-g++-4.8/test/BGL/test_Euler_operations.cpp:386:3: note: in instantiation of function template specialization 'test_Euler_operations<OpenMesh::PolyMesh_ArrayKernelT<OpenMesh::DefaultTraits> >' requested here
  test_Euler_operations<OMesh>();
  ^
```
and other errors.

**Comment by afabri** [comment]
When called outside of the namespace CGAL, calls of functions like `halfedges_around_face()` must be prefixed with `CGAL::` if applied to face graphs in a non-CGAL namespace.

**Comment by lrineau** [comment]
@afabri:
> When called outside of the namespace CGAL, calls of functions like `halfedges_around_face()` must be prefixed with `CGAL::` if applied on face graphs in a non-CGAL namespace. 

They are not supposed to be found by the argument-dependent lookup (ADL)?!

**Comment by afabri** [comment]
No, because they are generic, and not part of the set of functions that make a class model of the concept `FaceGraph`.

**Comment by gdamiand** [comment]
There errors are in master but the code is not tested.

My modification in this commit https://github.com/CGAL/cgal/pull/958/commits/5f7697c3afb2bce430da80baa4a71da71616d259 implies that now this code is tested but I am not gulty ;)

I think this bug is perhaps the same than this one https://github.com/CGAL/cgal/issues/942  ?

**Comment by gdamiand** [comment]
@afabri Do you plan to make the corrections ?
If not I will remove this commit https://github.com/CGAL/cgal/pull/958/commits/5f7697c3afb2bce430da80baa4a71da71616d259.

**Comment by afabri** [comment]
If you want I make the bug fix in a separate PR, but it is probably easier, if you just do it in this PR.

**Comment by lrineau** [comment]
@gdamiand Please add me and @sloriot as collaborator in your fork, here: https://github.com/gdamian/cgal/settings/collaboration

**Comment by lrineau** [comment]
I think the fix should be in this PR, but you be done by persons from @CGAL/geometryfactory .

**Comment by gdamiand** [comment]
@lrineau Done (and thanks to do the fix).

**Comment by lrineau** [comment]
Will be tested tonight.

**Comment by lrineau** [comment]
TODO for @lrineau: merge in CGAL-4.8 as well as in master.

File: BGL/test/BGL/test_Prefix.h [hunk 1]
```diff
@@ -226,25 +226,25 @@ struct Surface_fixture_2 {
     bool found;
     boost::tie(h, found) = halfedge(x, v, m);
     assert(found);
-    assert(! is_border(h,m));
+    assert(! CGAL::CGAL::is_border(h,m));
     f1 = face(h, m);
```
Comment (lrineau) [inline BGL/test/BGL/test_Prefix.h]: Compilation error, here.

**Comment by lrineau** [comment]
Now there is one assertion error, on all platforms with OpenMesh:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.9-Ic-13/BGL/TestReport_lrineau_Fedora-strict-ansi.gz
```
------------------------------------------------------------------
- ProgramOutput.test_Euler_operations.Fedora-strict-ansi
------------------------------------------------------------------
CGAL::get_default_random()::get_seed() = 1459587286
ImporterT: Face contains invalid vertex index
ImporterT: Face contains invalid vertex index
ImporterT: Face contains invalid vertex index
test_Euler_operations: /home/cgal_tester/build/src/cmake/platforms/Fedora-strict-ansi/test/BGL/test_Prefix.h:170: Surface_fixture_1<Graph>::Surface_fixture_1() [with Graph = OpenMesh::PolyMesh_ArrayKernelT<>]: Assertion `u != boost::graph_traits<Graph>::null_vertex()' failed.
./cgal_test_with_cmake: line 69: 31125 Aborted                 (core dumped) ./test_Euler_operations
```

Please run the testsuite locally.

**Comment by afabri** [comment]
I had tried to reproduce it with VC++, but  did not succeed.

**Comment by lrineau** [comment]
@gdamiand @afabri 

I have rebased the branch onto `releases/CGAL-4.8-branch`. Please use:
```
git checkout  BGL-bugfix-gdamiand
git fetch gdamiand # or `git fetch mine`, for Guillaume
git reset --hard gdamiand/BGL-bugfix-gdamiand # or `mine/...` for Guillaume
```
before any new commit.

**Comment by lrineau** [comment]
@afabri @gdamiand:
This PR will be tested alone in `integration` tonight.

If the test are good, tomorrow I will merge that `integration` into `releases/CGAL-4.8-branch`, with a fast-forward merge (that means that the value of `releases/CGAL-4.8-branch` will be set to the value of `integration`). That will allow to gain one day: no need to test `releases/CGAL-4.8-branch` by itself.

@ndouillet @sloriot: Please to not merge anything into `integration` today.

**Comment by lrineau** [comment]
It was successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.8-I-163.shtml

**Pull request merged**